### PR TITLE
Implementation of hooks

### PR DESCRIPTION
**Pull request by Tyrannas**

### CODE REVIEW

File: mlflow_client/hooks.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+     'run_started': [
+        {
+            'name': 'firstHook'
+            'url': 'http://sgaiapi/new_run
+        }
+        ...
+     ]
+     ...
+}
+"""
```
Comment (Tyrannas) [inline mlflow_client/hooks.py]: @gfournier Does this seem ok ?

File: experiments/main.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
     # backend = get_auto_backend()
 
     # backend = MLFlowBackend()
-    backend = LocalBackend()
+    backend = LocalBackend(hooks_uri="http://sgai-api/api/hooks")
```
Comment (gfournier) [inline experiments/main.py]: If it's a sample then `http://localhost/api/hooks`

File: mlflow_client/hooks.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class Hooks(Enum):
```
Comment (gfournier) [inline mlflow_client/hooks.py]: `Hook`

File: mlflow_client/hooks.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import requests
+
+"""
+Hooks template file:
+{
+     'run_started': [
+        {
+            'name': 'firstHook'
+            'url': 'http://sgaiapi/new_run
```
Comment (gfournier) [inline mlflow_client/hooks.py]: Use `localhost` for hosts in documentation

File: mlflow_client/hooks.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    # TODO: implement behaviour
+    return {}
+
+
+def send_hook(action: Hooks, experiment_name, run_id, url):
+    """
+    Create the hook object and send it to the specified URI
+    """
+    hook = {
```
Comment (gfournier) [inline mlflow_client/hooks.py:114]: Don't we use a more generic payload ?

```json
{
  "event": "start",
  "timestamp": "2020-03-30T23:45:56",
  "payload": {
    "experiment": "myExperiment",
    "run": "myRunId"
  }
```

File: mlflow_client/hooks.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        'action': action.name,
+        'timestamp': str(datetime.datetime.now()),
```
Comment (gfournier) [inline mlflow_client/hooks.py]: Format datetime as `'YYYY-MM-ddTHH:mm:ss'` in JSON, not sure that `str()` makes it the right way.

File: mlflow_client/backends.py [hunk 7]
```diff
@@ -134,11 +140,51 @@ def end_run(self):
...
+    """
+    def __init__(self):
+        pass
+
+    def __call__(self, cls):
+        class BackendWithHooks(cls):
+            """
+            SubClass of the decorated Backend that implements Hooks
+            """
+            def __init__(self, *args, **kwargs):
```
Comment (gfournier) [inline mlflow_client/backends.py]: Ok but add docstring for `hooks_uri`.

File: mlflow_client/backends.py [hunk 8]
```diff
@@ -134,11 +140,51 @@ def end_run(self):
...
+                # TODO: the problem is that hooks uri is not documented when creating a LocalBackend for example
+                uri = kwargs['hooks_uri'] if 'hooks_uri' in kwargs else None
+                self.hooks = resolve_hooks(uri)
+                super().__init__(*args, **kwargs)
+
+            def start_run(self, *args, **kwargs):
```
Comment (gfournier) [inline mlflow_client/backends.py]: I'm wondering if we can intercept any kind of method ?

**Review by gfournier — CHANGES_REQUESTED** [review]

File: mlflow_client/backends.py (hunk 7 above)
Comment (Tyrannas) [reply mlflow_client/backends.py]: That's what I mentionned before, by using a class decorator to add default behaviour, the 'hooks_uri' parameter cannot be documented by autocompletion when trying to do LocalBackend() for example

Even if I had docstring to the subclass

File: mlflow_client/backends.py (hunk 8 above)
Comment (gfournier) [reply mlflow_client/backends.py]: You can try this. We decorate each method we want to "hook" and pass the hook name as a decorator parameter.

```python
def log(s):
    def decorator(fn):
        def wrapper(self):
            print('In wrapper of %s' % self.name)
            print('[%s] before calling %s' % (s, fn.__name__))
            try:
                return fn(self)
            finally:
                print('[%s] done calling %s' % (s, fn.__name__))
        return wrapper
    return decorator

class C:
    def __init__(self):
        self.name = "test"
    @log("AAA")
    def printA(self):
        print("A")
    def printB(self):
        print("B")

C().printA()
C().printB()
```

**Review by gfournier — APPROVED** [review]

**Pull request merged**